src/connection.js

```diff
@@ -0,0 +1,282 @@
+									this.state = DO_NOTHING;
+									resolve();
+								}
+								return;
+							}
+
+							// if we aren't expecting the data to have a length (i.e. the shell command),
+							// then buffer immediately
+							if (this.opts.noLength) {
+								this.emit(dataEventName, `[${conn.connNum}] PUSHING REMAINING DATA INTO BUFFER AND SETTING STATE TO BUFFER_UNTIL_CLOSE`);
+								this.state = BUFFER_UNTIL_CLOSE;
+								return;
+							}
+
+							this.state = WAIT_FOR_NEW_DATA;
+							len = null; // we don't know the length yet
+							// purposely fall through
+
+						case WAIT_FOR_NEW_DATA:
+							// find how many bytes we are waiting for
+							if (len === null && buffer.length >= 4) {
+								len = parseInt(buffer.slice(0, 4), 16);
+								this.emit(dataEventName, `[${conn.connNum}] DETERMINING EXPECTED LENGTH...`);
+								isNaN(len) && (len = null);
+								buffer = buffer.slice(4);
+							}
+
+							// if there's no length, then let's fire the callback or wait until the socket closes
+							if (len === 0) {
+								this.emit(dataEventName, `[${conn.connNum}] NO EXPECTED LENGTH, FIRING CALLBACK`);
+								buffer = null;
+								len = null;
+								return resolve();
+							} else if (len === null) {
+								this.emit(dataEventName, `[${conn.connNum}] NO EXPECTED LENGTH`);
+								if (this.opts.bufferUntilClose) {
+									this.emit(dataEventName, `[${conn.connNum}] BUFFERING DATA UNTIL SOCKET CLOSE`);
+									this.state = BUFFER_UNTIL_CLOSE;
+								} else  {
+									buffer = null;
+									len = null;
+									this.state = WAIT_FOR_NEW_DATA;
+									resolve();
+								}
+								return;
+							}
+
+							this.emit(dataEventName, `[${conn.connNum}] EXPECTED LENGTH = ${len}`);
+							this.emit(dataEventName, `[${conn.connNum}] BUFFER LENGTH = ${buffer.length}`);
+
+							// do we have enough bytes?
+							if (buffer.length >= len) {
+								// yup
+								let result = buffer.slice(0, len);
+								buffer = buffer.slice(len);
+								this.emit(dataEventName, `[${conn.connNum}] SUCCESS AND JUST THE RIGHT AMOUNT OF BYTES (${len}) WITH ${buffer.length} BYTES LEFT`);
+								if (this.opts.bufferUntilClose) {
+									this.state = BUFFER_UNTIL_CLOSE;
+								} else {
+									this.state = WAIT_FOR_NEW_DATA;
+									len = null;
+									buffer = null;
+									resolve(result);
+								}
+							} else {
+								// we need more data!
+								this.emit(dataEventName, `[${conn.connNum}] WAITING FOR MORE DATA`);
+							}
+							return;
+
+						case BUFFER_UNTIL_CLOSE:
+							// we've already added data to the buffer
+							return;
+					}
+				}
+			});
+
+			socket.on('end', () => {
+				this.emit(dataEventName, `[${this.connNum}] SOCKET CLOSED BY SERVER, (${buffer} && ${buffer && buffer.length})`);
+				this.end();
+
+				if (buffer && buffer.length) {
+					resolve(buffer);
+					buffer = null;
+				} else {
+					resolve();
```

Comment:
Do we need this `resolve()`? It wasn't in the original implementation: https://github.com/appcelerator/titanium_mobile/blob/master/node_modules/titanium-sdk/lib/adb.js#L248-L255.